### PR TITLE
disallow typedesc in arrays & move existing checks to `types.typeAllowedAux`

### LINKED ISSUE #9932: Error: internal error: expr(skType); unknown symbol
```
let a = [ int ]
```

```ico@platdoos:/tmp$ nim c tt.c
tt.c(1, 11) Error: internal error: expr(skType); unknown symbol
No stack traceback available
To create a stacktrace, rerun compilation with ./koch temp c <file>
```

Related to #7331?

### PR DESCRIPTION
**Pull request by nc-x**
Closes https://github.com/nim-lang/Nim/issues/9932

### CODE REVIEW

**Comment by Araq** [comment]
But this is arbitrary, instead it should call ``types.typeAllowed`` or ``types.typeAllowed`` should be patched to disallow it.

File: compiler/sem.nim [hunk 1]
```diff
@@ -233,13 +233,18 @@ proc typeAllowedCheck(conf: ConfigRef; info: TLineInfo; typ: PType; kind: TSymKi
                       flags: TTypeAllowedFlags = {}) =
   let t = typeAllowed(typ, kind, flags)
   if t != nil:
+    var err: string
     if t == typ:
-      localError(conf, info, "invalid type: '" & typeToString(typ) &
-        "' for " & substr($kind, 2).toLowerAscii)
+      err = "invalid type: '" & typeToString(typ) &
+        "' for " & substr($kind, 2).toLowerAscii
```
Comment (timotheecour) [inline compiler/sem.nim]: `ast.toHumanStr`

Comment (timotheecour) [inline compiler/sem.nim]: use `%`, cleaner/more readable

File: compiler/sem.nim [hunk 2]
```diff
@@ -233,13 +233,18 @@ proc typeAllowedCheck(conf: ConfigRef; info: TLineInfo; typ: PType; kind: TSymKi
...
+      if kind in {skVar, skLet, skConst} and taIsTemplateOrMacro in flags:
+        err &= ". Did you mean to call the $1 with '()'?" % [
+                               if typ.owner.kind == skMacro: "macro" else: "template"]
```
Comment (timotheecour) [inline compiler/sem.nim]: `toHumanStr`

File: compiler/sem.nim [hunk 3]
```diff
@@ -233,13 +233,18 @@ proc typeAllowedCheck(conf: ConfigRef; info: TLineInfo; typ: PType; kind: TSymKi
...
     else:
-      localError(conf, info, "invalid type: '" & typeToString(t) &
+      err = "invalid type: '" & typeToString(t) &
         "' in this context: '" & typeToString(typ) &
-        "' for " & substr($kind, 2).toLowerAscii)
+        "' for " & substr($kind, 2).toLowerAscii
```
Comment (timotheecour) [inline compiler/sem.nim]: toHumanStr

File: compiler/types.nim [hunk 4]
```diff
@@ -1307,18 +1309,24 @@ proc typeAllowedAux(marker: var IntSet, typ: PType, kind: TSymKind,
...
+      if result.isNil and t[0] != nil:
+        result = typeAllowedAux(marker, t[0], skResult, flags)
   of tyTypeDesc:
-    # XXX: This is still a horrible idea...
-    result = nil
+    if kind in {skVar, skLet, skConst} and taProcContextIsNotMacro in flags:
+      result = t
+    else:
+      # XXX: This is still a horrible idea...
+      result = nil
```
Comment (timotheecour) [inline compiler/types.nim:1329]: can it still happen now?

File: compiler/types.nim [hunk 5]
```diff
@@ -1307,18 +1309,24 @@ proc typeAllowedAux(marker: var IntSet, typ: PType, kind: TSymKind,
...
       result = t
-    let f = if kind in {skProc, skFunc}: flags+{taNoUntyped} else: flags
-    for i in 1..<t.len:
-      if result != nil: break
-      result = typeAllowedAux(marker, t[i], skParam, f-{taIsOpenArray})
-    if result.isNil and t[0] != nil:
-      result = typeAllowedAux(marker, t[0], skResult, flags)
+    else:
+      if isInlineIterator(typ) and kind in {skVar, skLet, skConst, skParam, skResult}:
+        # only closure iterators my be assigned to anything.
```
Comment (timotheecour) [inline compiler/types.nim]: s/my/may/

File: compiler/types.nim (hunk 4 above)
Comment (nc-x) [reply compiler/types.nim:1329]: This was existing code, so i don't know anything about this.
While reviewing, use the settings icon to **Hide Whitespace Changes** in order to see which code this PR changes.

**Comment by nc-x** [comment]
All tests pass. Ready for review. @Araq 

Edit: Why does `[skip ci]` not work?

**Comment by timotheecour** [comment]
> Edit: Why does [skip ci] not work?

you can upvote the issue I filed: https://github.com/microsoft/azure-pipelines-agent/issues/2944
the workaround is to implement something like https://github.com/nim-lang/Nim/pull/13556 `allow [nimDocOnly] in commit msg` althouhg there are other ways to implement it

**Comment by Araq** [comment]
Do not use ``[skip ci]``, it's better to wait for the CIs, they are getting faster over time.

**Comment by nc-x** [comment]
AFAICT, the currently used CI services azure, github actions as well as builds.sr.ht do not support [ci skip], so we cannot use it anyways.

**Comment by timotheecour** [comment]
> Do not use [skip ci], it's better to wait for the CIs, they are getting faster over time.

it's still useful for in-progress work, but yes when it's getting finalized CI is almost always needed.
Another typical example (while still in development)is: your CI is almost finished completed, but a push (say to fix a typo) would restart all jobs. sad.

**Comment by Araq** [comment]
Superb work, many thanks!

**Pull request merged**